config/config.go

```diff
@@ -66,56 +110,83 @@ func ReadConfigFromFile() (ShipperConfig, error) {
+	cfg, err := config.NewConfigFrom(mapCfg)
 	if err != nil {
-		return ShipperConfig{}, err
+		return ShipperRootConfig{}, fmt.Errorf("error reading in raw map config: %w", err)
 	}
-	return ReadConfigFromJSON(string(jsonConfig))
-}
 
-// ReadConfigFromJSON reads the event in from a JSON config. I believe @blakerouse told me
-// that the V2 controller will send events via JSON, but I could be wrong.
-func ReadConfigFromJSON(raw string) (ShipperConfig, error) {
-	rawCfg, err := json.NewConfig([]byte(raw))
+	// We should merge config overwrites here from the -E flag,
+	// but they seem to step on the elasticsearch config,
+	// so for now, don't.
+
+	err = cfg.Unpack(&cfgObject)
 	if err != nil {
-		return ShipperConfig{}, fmt.Errorf("error parsing string config: %w", err)
+		return ShipperRootConfig{}, fmt.Errorf("error unpacking shipper config: %w", err)
 	}
 
-	unpacker := func(cfg *ShipperConfig) error {
-		return rawCfg.Unpack(cfg)
+	cfgObject.Shipper.Server = grpcConfig
+
+	// output config is at the "root" level, so we need to unpack those manually
+	if cfgObject.Type == esKey {
+		err = cfg.Unpack(&cfgObject.Shipper.Output.Elasticsearch)
+		if err != nil {
+			return ShipperRootConfig{}, fmt.Errorf("error reading elasticsearch output: %w", err)
+		}
+	} else if cfgObject.Type == kafaKey {
+		err = cfg.Unpack(&cfgObject.Shipper.Output.Kafka)
+		if err != nil {
+			return ShipperRootConfig{}, fmt.Errorf("error reading elasticsearch output: %w", err)
+		}
+	} else if cfgObject.Type == consoleKey {
+		err = cfg.Unpack(&cfgObject.Shipper.Output.Console)
+		if err != nil {
+			return ShipperRootConfig{}, fmt.Errorf("error reading console output: %w", err)
+		}
+	} else {
+		return ShipperRootConfig{}, fmt.Errorf("error, could not find output for output key '%s'", cfgObject.Type)
 	}
-
-	return readConfig(unpacker)
+	return cfgObject, nil
 }
 
-type rawUnpacker func(cfg *ShipperConfig) error
+type rawUnpacker func(cfg *ShipperRootConfig) error
 
-func readConfig(unpacker rawUnpacker) (config ShipperConfig, err error) {
+func readConfig(unpacker rawUnpacker) (config ShipperRootConfig, err error) {
 	// systemd environment will send us to stdout environment, which we want
-	config = ShipperConfig{
-		Log:     logp.DefaultConfig(logp.SystemdEnvironment),
-		Monitor: monitoring.DefaultConfig(),
-		Queue:   queue.DefaultConfig(),
-		Server:  server.DefaultConfig(),
-		Output:  output.DefaultConfig(),
-	}
-
+	config = DefaultConfig()
 	err = unpacker(&config)
 	if err != nil {
 		return config, fmt.Errorf("error unpacking shipper config: %w", err)
 	}
 
-	// otherwise the logging configuration is just ignored
-	err = logp.Configure(config.Log)
-	if err != nil {
-		return config, fmt.Errorf("error configuring the logger: %w", err)
-	}
-
 	return config, nil
 }
+
+// ZapFromUnitLogLevel converts the log level used by the units API, to the one used by the logger
+func ZapFromUnitLogLevel(level client.UnitLogLevel) zapcore.Level {
```

Comment:
I recognize this from Beats (or maybe Agent). This should probably live in elastic-agent-client.

Reply:
Do we just want a straight `ZapFromUnitLogLevel` helper function, or should we modify the main `unit.Expected()` to return a zap level object?

Reply:
I think we want the helper function. The client shouldn't assume the logging package in use, but it can provide helpers for the most common case.

Reply:
I don't feel super strongly about this if you don't want to do it.

Reply:
Yah, kinda tempted to make a follow-up issue for this, since the PR itself has been delayed long enough....